src/java.base/share/classes/jdk/internal/lang/stable/StableFieldUpdater.java

```diff
@@ -0,0 +1,358 @@
+                                                    long zeroReplacement) implements ToLongFunction<T> {
+
+        @ForceInline
+        @Override
+        public long applyAsLong(T t) {
+            checkType(holderType, t);
+            // Plain semantics suffice here as we are not dealing with a reference (for
+            // a reference, the internal state initialization can be reordered with
+            // other store ops). JLS (24) 17.4 states that 64-bit fields tear under
+            // plain memory semantics and this VM is not 64-bit.
+            long v = UNSAFE.getLongOpaque(t, offset);
+            if (v == 0) {
+                // StableUtil.preventReentry(t);
+                synchronized (t) {
+                    v = UNSAFE.getLongAcquire(t, offset);
+                    if (v == 0) {
+                        v = underlying.applyAsLong(t);
+                        if (v == 0) {
+                            v = zeroReplacement;
+                        }
+                        UNSAFE.putLongRelease(t, offset, v);
+                    }
+                }
+            }
+            return v;
+        }
+    }
+
+    // Static support functions
+
+    @ForceInline
+    private static void checkType(Class<?> holderType, Object t) {
+        if (!holderType.isInstance(t)) {
+            throw new IllegalArgumentException("The provided t is not an instance of " + holderType);
+        }
+    }
+
+    private static long offset(Class<?> holderType,
+                               String fieldName,
+                               Class<?> fieldType,
+                               Class<?> caller) {
+        final Field field;
+        try {
+            field = findField(holderType, fieldName);
+            int modifiers = field.getModifiers();
+            if (Modifier.isFinal(modifiers)) {
+                throw illegalField("non final fields", field);
+            }
+            sun.reflect.misc.ReflectUtil.ensureMemberAccess(
+                    caller, holderType, null, modifiers);
+            if (field.getType() != fieldType) {
+                throw illegalField("fields of type '" + fieldType + "'", field);
+            }
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new IllegalArgumentException(e);
+        }
+        return UNSAFE.objectFieldOffset(field);
+    }
+
+    private static IllegalArgumentException illegalField(String msg, Field field) {
+        return new IllegalArgumentException("Only " + msg + " are supported. The provided field is '" + field + "'");
+    }
+
+    private static Field findField(Class<?> holderType, String fieldName) throws NoSuchFieldException {
+        if (holderType.equals(Object.class)) {
+            throw new NoSuchFieldException("'" + fieldName + "' in '" + holderType + "'");
+        }
+        final Field[] fields = holderType.getDeclaredFields();
+        for (Field f : fields) {
```

Comment:
```suggestion
        for (Field f : holderType.getDeclaredFields()) {
```
The local variable fields is only used once and can be simplified